iseq.c

```diff
@@ -317,19 +318,69 @@ rb_iseq_each_value(const rb_iseq_t *iseq, iseq_value_itr_t * func, void *data)
 {
     unsigned int size;
     VALUE *code;
-    size_t n;
-    rb_vm_insns_translator_t *const translator =
-#if OPT_DIRECT_THREADED_CODE || OPT_CALL_THREADED_CODE
-        (FL_TEST((VALUE)iseq, ISEQ_TRANSLATED)) ? rb_vm_insn_addr2insn2 :
-#endif
-        rb_vm_insn_null_translator;
     const struct rb_iseq_constant_body *const body = ISEQ_BODY(iseq);
 
     size = body->iseq_size;
     code = body->iseq_encoded;
 
-    for (n = 0; n < size;) {
-	n += iseq_extract_values(code, n, func, data, translator);
+    union iseq_inline_storage_entry *is_entries = body->is_entries;
+
+    // IVC and ICVARC entries
+    for (unsigned int i = 0; i < body->ivc_size; i++, is_entries++) {
+        IVC ivc = (IVC)is_entries;
+        if (ivc->entry) {
+            if (RB_TYPE_P(ivc->entry->class_value, T_NONE)) {
+                rb_bug("!! %u", ivc->entry->index);
+            }
+
+            VALUE nv = func(data, ivc->entry->class_value);
+            if (ivc->entry->class_value != nv) {
+                ivc->entry->class_value = nv;
+            }
+        }
+    }
+
+    // ISE entries
+    for (unsigned int i = 0; i < body->ise_size; i++, is_entries++) {
+        union iseq_inline_storage_entry *const is = (union iseq_inline_storage_entry *)is_entries;
+        if (is->once.value) {
+            VALUE nv = func(data, is->once.value);
+            if (is->once.value != nv) {
+                is->once.value = nv;
+            }
+        }
+    }
+
+    // IC Entries
+    for (unsigned int i = 0; i < body->ic_size; i++, is_entries++) {
+        IC ic = (IC)is_entries;
+        if (ic->entry) {
+            VALUE nv = func(data, (VALUE)ic->entry);
+            if ((VALUE)ic->entry != nv) {
+                ic->entry = (void *)nv;
+            }
+        }
+    }
+
+    // Embedded VALUEs
+    for (unsigned int i = 0; i < ISEQ_MBITS_BUFLEN(size); i++) {
+        iseq_bits_t bits = body->mark_offset_bits[i];
+        if (bits) {
+            unsigned int count = 0;
+
+            while(bits) {
+                if (bits & 0x1) {
```

Comment:
I tried using `ffs` but it's turning out to be a pain because you can't right shift by 64.  So we'd need to add a special case if just the top bit is set.  I think we should investigate using `ffs`, but I want to merge this as-is for now.

Reply:
Hum, not sure I follow what the problem is. If only the first bit is set, `ffs` (or `ffl` more likely) return `64`, but the shift should always be `bits >>= ffs(bits) - 1` (unless ffs returns `0`).

Anyway no big deal. I'd like to try backporting this patch on top of 3.1 though.

Reply:
This post about [iterating over set bits quickly](https://lemire.me/blog/2018/02/21/iterating-over-set-bits-quickly/) could be useful here, it uses ``__builtin_ctzl``.

Reply:
> Hum, not sure I follow what the problem is. If only the first bit is set, ffs (or ffl more likely) return 64, but the shift should always be bits >>= ffs(bits) - 1 (unless ffs returns 0).

`ffs` returns the nth bit and starts at a 1 index.  So `ffs(0x1) => 1`.  If we did `bits >>= ffs(bits) - 1`, then it would never shift anything in the case of `0x1`, so we have to do `bits >>= ffs(bits)`.  `ffs` of `0x8000000000000000` returns 64, so the right shift doesn't work.

We can probably use `__builtin_ctzl`, but I think we should see if this iteration is really a bottleneck at the moment.